Fix dataresend

MCLiii opened the pull request.

jrvollmer commented on src/routes/api.js:
```diff
@@ -238,16 +239,41 @@ function openSocket() {
+   while (startIndex !== -1 && endIndex !== -1) {
+     console.log("start index " + startIndex, "end index " + endIndex)
+     // Extract a complete data packet, this is different from chase car dashboard since we want to include the <br></br>
+     const compete_packet = Buffer.from(partialData.slice(startIndex, endIndex + 5),'latin1');
```
```suggestion
     const complete_packet = Buffer.from(partialData.slice(startIndex, endIndex + 5),'latin1');
```
nit

jrvollmer commented on src/routes/api.js:
```diff
@@ -238,16 +239,41 @@ function openSocket() {
+        unpackData(data);
+     } else {
+       console.warn("ERROR: Bad packet length ------------------------------------");
+     }
```
```suggestion
     if (complete_packet.length === (bytesPerPacket + 9)) {
        broadcastData(complete_packet);
     } else {
       console.warn("ERROR: Bad packet length ------------------------------------");
     }
```

jrvollmer commented on src/routes/api.js:
```diff
@@ -238,16 +239,41 @@ function openSocket() {
+     startIndex = partialData.indexOf("<bl>");
+     endIndex = partialData.indexOf("</bl>");
+   }
+   // If the remaining data is longer than the expected packet length,
+   // there might be an incomplete packet, so log a warning.
+   if (partialData.length >= bytesPerPacket) {
+     console.warn("ERROR: Incomplete or malformed packet ------------------------------------");
+     partialData = '';
+   }
```
I'm not 100% sure how I feel about this. @MCLiii, Is the following the general idea?
```
TCP buffer 1 (fills up bc we resent a lot of data at once): <bl>... (a bunch of packets) ...</bl><bl> (a partial packet)

TCP buffer 2 (not yet handled by the "data" event handler): (rest of the partial packet from the end of TCP buffer 1) </bl>...
```
Then, this check is meant to allow the partial packets at the end and beginning of the two buffers to be merged when the "data" event handler receives TCP buffer 2?

**If so, correct me if I'm wrong, but I think we should make this check the following instead: `partialData.length >= bytesPerPacket + 8`. That will check if the  packet is anything more than  `<bl> (packet contents) </bl` (just missing the ending '>')**

MCLiii replied:
You are right, it should be bytesPerPacket+8

jrvollmer commented on src/routes/api.js:
```diff
@@ -238,16 +239,41 @@ function openSocket() {
-  });
+ // Data received listener
+ let partialData = '';
+
```
```suggestion

 let partialData = '';
 // Data received listener
```
nit

jrvollmer commented on src/routes/api.js:
Let's add one space to the indentation of the added lines

jrvollmer commented on src/routes/api.js:
```diff
@@ -238,16 +239,41 @@ function openSocket() {
+   // Combine the new data with any previously received partial data
+   partialData += data.toString('latin1');
+
+   let startIndex = partialData.indexOf("<bl>");
+   let endIndex = partialData.indexOf("</bl>");
+
+   while (startIndex !== -1 && endIndex !== -1) {
```
Let's add a check here to see if `partialData.length === bytesPerPacket` (and maybe that `startIndex !== 0` and/or `endIndex !== bytesPerPacket + 4`?). If so, pass it on as is. This will make the API backwards compatible by allowing packets without \<bl\>\</bl\> to pass through to the chase car dashboards

jrvollmer reviewed (changes requested):
@MCLiii, I'll make the changes and add them to the PR. **Please review my comments/suggestions and the changes I make to make sure they all look good**

MCLiii commented on src/routes/api.js:
```diff
@@ -238,16 +239,41 @@ function openSocket() {
+   }
+   // If the remaining data is longer than the expected packet length,
+   // there might be an incomplete packet, so log a warning.
+   if (partialData.length >= bytesPerPacket) {
```
```suggestion
   if (partialData.length >= bytesPerPacket + 8) {

```

MCLiii committed:
Create server to interface with solar car clients

Listen for connections from the solar car on port 4005 to enable the
solar car to determine whether or not it can establish a TCP connection
with this server.

jrvollmer commented on src/routes/api.js:
```diff
@@ -268,15 +306,50 @@ function openSocket() {
+
+    let startIndex = partialData.indexOf("<bl>");
+    let endIndex = partialData.indexOf("</bl>");
+
+    // Check if the packet is the size of a nominal payload without <bl></bl>
+    // If so, assume it is a nominal packet without <bl></bl> and broadcast it as is
+    if (partialData.length === bytesPerPacket) {
+      broadcastData(Buffer.from(partialData,'latin1'));
+      partialData = '';
     } else {
-      readline.pause();
-      console.warn("\nERROR: Bad packet length ------------------------------------");
-      readline.prompt(true);
+      while (startIndex !== -1 && endIndex !== -1) {
+        // Extract a complete data packet, this is different from chase car dashboard since we want to include the <br></br>
+        const complete_packet = Buffer.from(partialData.slice(startIndex, endIndex + 5),'latin1');
+
+        // Process the complete data packet
+        if (complete_packet.length === (bytesPerPacket + 9)) {
+          broadcastData(complete_packet);
+        } else {
+          readline.pause();
+          console.warn("ERROR: Bad packet length ------------------------------------");
+          readline.prompt(true);
+        }
+        // Update the partial data to exclude the processed packet
+        partialData = partialData.substring(endIndex + 5);
+        // Search for the next complete data packet
+        startIndex = partialData.indexOf("<bl>");
+        endIndex = partialData.indexOf("</bl>");
+      }
+
+      // If the remaining data is longer than the expected packet length,
+      // there might be an incomplete packet, so log a warning.
+      if (partialData.length >= bytesPerPacket + 8) {
+        readline.pause();
+        console.warn("ERROR: Incomplete or malformed packet ------------------------------------");
+        readline.prompt(true);
+        partialData = '';
+      }
```
@MCLiii, I just realized that it really doesn't pay to do this here. Since we'll be parsing the incoming packets on each chase-car-dashboard instance, we don't need to do it here. Also, we would need to parse them for packets from both TCP and the VPS, so it's more upkeep here.

jrvollmer replied:
Okay, thinking about it more, maybe there's merit to checking the validity of packets here and then sending individual complete packets to the dashboards. However, it's also less efficient to send individual packets rather than streams of multiple packets. I'm going to stash these changes and open a draft PR for now

jrvollmer commented:
Closing because #3 has been merged
